rpc/rpc/v1_0/handlers.py

```diff
@@ -0,0 +1,84 @@
+"""Message handlers for DIDComm RPC v1.0."""
+
+import json
+from aries_cloudagent.messaging.base_handler import (
+    BaseResponder,
+    BaseHandler,
+    RequestContext,
+)
+from aries_cloudagent.storage.base import BaseStorage
+from aries_cloudagent.storage.record import StorageRecord
+
+from rpc.v1_0.messages import DRPCRequestMessage, DRPCResponseMessage
+from rpc.v1_0.models import DRPCRecord
+
+
+class DRPCRequestHandler(BaseHandler):
+    """Handler class for DRPCRequestMessage."""
+
+    async def handle(self, context: RequestContext, responder: BaseResponder):
+        """Message handler implementation."""
+
+        self._logger.debug("DRPCRequestHandler called with context %s", context)
+        assert isinstance(context.message, DRPCRequestMessage)
+
+        self._logger.info("Received RPC request: %s", context.message.request)
+
+        connection_id = context.connection_record.connection_id
+        thread_id = context.message._id
+
+        async with context.session() as session:
+            request_record = DRPCRecord(
+                request=context.message.request, state=DRPCRecord.STATE_REQUEST_RECEIVED
+            )
+            storage = session.inject(BaseStorage)
+            record = StorageRecord(
+                type=DRPCRecord.RECORD_TYPE,
+                value=json.dumps(request_record.serialize()),
+                tags={
+                    "connection_id": connection_id,
+                    "thread_id": thread_id,
+                },
+            )
+            await storage.add_record(record)
+
+        await context.profile.notify(
```

Comment:
Is there a event bus handler for this? I'm not seeing where it's registered. Is this doing anything or supposed to be?

I don't have it open in my code editor so might just be missing something

Reply:
There isn't. I wasn't sure where to register the events

Reply:
If you want to register the events you can do it like this in your routes.py file.

![image](https://github.com/hyperledger/aries-acapy-plugins/assets/31809382/8cfeb6c9-56d2-4b6b-91f4-b887e88eb8f3)

The naming is important. The agent looks in all the routes.py files for this method when intializing the event bus. You can see them getting registered with log-level set to debug.
 
If it doesn't need to do anything with this message than you don't need this though.

Reply:
I definitely want controllers to be notified of incoming messages for this protocol so this helps.

Reply:
Will the event bus result in web hook callbacks? That's really what I'm looking to accomplish

Reply:
Oh. I'm not an expert on this but I don't think so. The event bus is for messages within one agent. Which is handy for decoupling components.
Do these agents have a connection?
If you want to notify another agent then I think you need to use the `BaseResponder --> aries_cloudagent/messaging/responder.py`. You can inject it and then notify another agents handler.

Ian or Daniel might be able to help better though. I haven't done a lot of stuff in this area.

Reply:
This would be in the case of an Agent notifying its own controller(s) through the webhooks. Perhaps @ianco or @dbluhm could weigh in.

Reply:
The webhook emitter is itself an event listener listening on a specific set of events:

https://github.com/hyperledger/aries-cloudagent-python/blob/main/aries_cloudagent/admin/server.py#L47-L59

Reply:
So that means I'll have to make a corresponding change in ACA-Py directly, since plugins up to this point haven't had the requirement of calling back to controllers. Is that correct? Do we want to generalize this in the plugins or should I just add this to the current list of events in ACA-Py for now?

Reply:
I think what I would recommend is to emit two events: 
- One for `drpc::request::received`; this could be consumed by plugin based rpc handlers
- One for `acapy::webhook::drpc-request`; this would be consumed by the admin server webhook listener to then in turn be consumed by controller based rpc handlers

Reply:
Ok that makes sense. I'll emit a second event which we'll register with the Admin event listeners

Reply:
The admin server is already listening for `acapy::webhook::(.*)`, fortunately, so no changes should be required upstream. It will turn it into a webhook to `/topic/drpc-request` (maybe we should do an underscore instead of a hyphen since I think that's what is used in other webhook topics).

Reply:
Tbh we don't really need the first event emitted. The DRPC handlers don't listen for those events.

Reply:
No events inside of this plugin, sure. But I like the idea of being able to define handlers from another plugin so I think the first event would be handy to keep around. And listening on `acapy::webhook::drpc-request` is doable but feels odd. But I don't feel strongly on that point.

Reply:
I meant to have that in snake case. I can add the second emitter back as well.